### PR TITLE
[Writing Tools] Reverting a list transformation leaves a single errant bullet

### PR DESCRIPTION
**Pull request by rr-codes**
#### c35d1fad75d58b9de8042a06f0fc21a9d411544e
<pre>
[Writing Tools] Reverting a list transformation leaves a single errant bullet
<a href="https://bugs.webkit.org/show_bug.cgi?id=275336">https://bugs.webkit.org/show_bug.cgi?id=275336</a>
<a href="https://rdar.apple.com/126139492">rdar://126139492</a>

Reviewed by Wenson Hsieh and Tim Horton.

Previously, we were undoing and restoring the text by creating and storing document fragments
and replacing the current selection. However, this approach is flawed for cases where lists
or table elements are used, since selecting all contents will exclude the first list bullet.

Fix by changing the implementation to instead just unapply and reapply the editing command itself,
which does not rely on selection.

Also, slightly refactor all the stateful instance variables into a single state struct.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/ReplaceSelectionCommand.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::willBeginTextReplacementSession):
(WebCore::Page::didBeginTextReplacementSession):
(WebCore::Page::textReplacementSessionDidReceiveReplacements):
(WebCore::Page::textReplacementSessionDidUpdateStateForReplacement):
(WebCore::Page::didEndTextReplacementSession):
(WebCore::Page::textReplacementSessionDidReceiveTextWithReplacementRange):
(WebCore::Page::contextRangeForSessionWithID const):
(WebCore::Page::textReplacementSessionDidReceiveEditAction):
* Source/WebCore/page/Page.h:
(WebCore::Page::unifiedTextReplacementController const): Deleted.
* Source/WebCore/page/unified-text-replacement/UnifiedTextReplacementController.h:
* Source/WebCore/page/unified-text-replacement/UnifiedTextReplacementController.mm:
(WebCore::UnifiedTextReplacementController::willBeginTextReplacementSession):
(WebCore::UnifiedTextReplacementController::textReplacementSessionDidReceiveReplacements):
(WebCore::UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement):
(WebCore::UnifiedTextReplacementController::textReplacementSessionDidReceiveTextWithReplacementRange):
(WebCore::UnifiedTextReplacementController::textReplacementSessionDidReceiveEditAction&lt;UnifiedTextReplacement::Session::ReplacementType::PlainText&gt;):
(WebCore::UnifiedTextReplacementController::textReplacementSessionDidReceiveEditAction&lt;UnifiedTextReplacement::Session::ReplacementType::RichText&gt;):
(WebCore::UnifiedTextReplacementController::textReplacementSessionDidReceiveEditAction):
(WebCore::UnifiedTextReplacementController::didEndTextReplacementSession&lt;UnifiedTextReplacement::Session::ReplacementType::PlainText&gt;):
(WebCore::UnifiedTextReplacementController::didEndTextReplacementSession):
(WebCore::UnifiedTextReplacementController::updateStateForSelectedReplacementIfNeeded):
(WebCore::UnifiedTextReplacementController::contextRangeForSessionWithID const):
(WebCore::UnifiedTextReplacementController::stateForSession):
(WebCore::UnifiedTextReplacementController::replaceContentsOfRangeInSessionInternal):
(WebCore::UnifiedTextReplacementController::replaceContentsOfRangeInSession):
* Source/WebKit/WebProcess/WebPage/Cocoa/TextIndicatorStyleController.mm:
(WebKit::TextIndicatorStyleController::contextRangeForSessionWithID const):

Canonical link: <a href="https://commits.webkit.org/279937@main">https://commits.webkit.org/279937@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/7d864149d71ec6fbe7b66ffc0f065f2970ae5f79

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54986 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34443 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7579 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/58264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5717 "Built successfully") 
| [❌ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57286 "Failed to checkout and rebase branch from PR 29692") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41993 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5749 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/58264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/3890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57081 "Failed to checkout and rebase branch from PR 29692") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/32529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/47640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/58264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/29316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/4984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3858 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/5205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59855 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30253 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/5362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/51953 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31383 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/47716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/51394 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32404 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8140 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31174 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/editing/ReplaceSelectionCommand.h [hunk 1]
```diff
@@ -27,11 +27,11 @@
 #pragma once
 
 #include "CompositeEditCommand.h"
+#include "DocumentFragment.h"
```
Comment (whsieh) [inline Source/WebCore/editing/ReplaceSelectionCommand.h:30]: Is this change necessary?

File: Source/WebCore/page/unified-text-replacement/UnifiedTextReplacementController.h [hunk 2]
```diff
@@ -64,9 +64,37 @@ class UnifiedTextReplacementController final {
 
     void updateStateForSelectedReplacementIfNeeded();
 
-    WEBCORE_EXPORT std::optional<SimpleRange> contextRangeForSessionWithID(const UnifiedTextReplacement::Session::ID&) const;
+    // FIXME: Remove this method when possible.
```
Comment (whsieh) [inline Source/WebCore/page/unified-text-replacement/UnifiedTextReplacementController.h]: Nit - can we elaborate on this FIXME? (It's unclear what `possible` entails)

**Review by hortont424 — APPROVED** [review]

File: Source/WebCore/page/unified-text-replacement/UnifiedTextReplacementController.mm [hunk 3]
```diff
@@ -141,19 +133,36 @@
...
+        return;
+    }
+
+    ASSERT(!m_states.contains(session->identifier));
+
+    auto liveRange = createLiveRange(*contextRange);
+
+    switch (session->replacementType) {
+    case UnifiedTextReplacement::Session::ReplacementType::PlainText:
+        m_states.set(session->identifier, PlainTextState { liveRange , 0 });
```
Comment (whsieh) [inline Source/WebCore/page/unified-text-replacement/UnifiedTextReplacementController.mm]: Nit -
```suggestion
        m_states.set(session->identifier, PlainTextState { liveRange, 0 });
```

File: Source/WebCore/page/unified-text-replacement/UnifiedTextReplacementController.mm [hunk 4]
```diff
@@ -174,79 +183,75 @@
...
 
     RefPtr document = this->document();
     if (!document) {
         ASSERT_NOT_REACHED();
         return;
     }
 
-    auto sessionRange = contextRangeForSessionWithID(session.identifier);
-    if (!sessionRange) {
+    auto* state = stateForSession<UnifiedTextReplacement::Session::ReplacementType::PlainText>(session);
```
Comment (whsieh) [inline Source/WebCore/page/unified-text-replacement/UnifiedTextReplacementController.mm]: Hm... it's probably okay as long as there's no reentrancy, but leaving raw pointers on the stack like this seems a bit counter to smart pointer deployment in WebKit.

Maybe we can use `CheckedPtr` instead?

**Review by whsieh — APPROVED** [review]

File: Source/WebCore/editing/ReplaceSelectionCommand.h (hunk 1 above)
Comment (rr-codes) [reply Source/WebCore/editing/ReplaceSelectionCommand.h:30]: before I added it, it was complaining about DocumentFragment being an incomplete type. I can try removing it and re-building to verify

File: Source/WebCore/page/unified-text-replacement/UnifiedTextReplacementController.mm (hunk 4 above)
Comment (rr-codes) [reply Source/WebCore/page/unified-text-replacement/UnifiedTextReplacementController.mm]: when I try that, I get
```
/Users/richardrobinson/safari/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/CheckedPtr.h:179:18: error: no member named 'incrementPtrCount' in 'WebCore::UnifiedTextReplacementController::PlainTextState'
```

Comment (rr-codes) [reply Source/WebCore/page/unified-text-replacement/UnifiedTextReplacementController.mm]: oh, should I conform those types to `CanMakeCheckedPtr`?

Comment (rr-codes) [reply Source/WebCore/page/unified-text-replacement/UnifiedTextReplacementController.mm]: oh I guess I would also have to define a constructor, and also WTF_STRUCT_OVERRIDE_DELETE_FOR_CHECKED_PTR

```
    struct RichTextState : CanMakeCheckedPtr<RichTextState> {
        WTF_MAKE_STRUCT_FAST_ALLOCATED;
        WTF_STRUCT_OVERRIDE_DELETE_FOR_CHECKED_PTR(RichTextState);

        RichTextState(const Ref<Range>& contextRange, const Vector<Ref<ReplaceSelectionCommand>>& commands)
            : contextRange(contextRange),
              commands(commands)
        {
        }

        Ref<Range> contextRange;
        Vector<Ref<ReplaceSelectionCommand>> commands;
    };
```
